pkg/api/steve/collections/rbac.go

```diff
@@ -0,0 +1,315 @@
+package collections
+
+import (
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/rancher/rancher/pkg/controllers/managementagent/nslabels"
+	pkgrbac "github.com/rancher/rancher/pkg/rbac"
+	corecontrollers "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	rbaccontrollers "github.com/rancher/wrangler/pkg/generated/controllers/rbac/v1"
+	"github.com/sirupsen/logrus"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	suffix                     = "-collections"
+	clusterRoleOwnerAnnotation = "authz.cluster.cattle.io/clusterrole-owner"
+)
+
+var ErrNotReady = errors.New("APIs have not been synced yet")
+
+type rbacHandler struct {
+	namespaceCache    corecontrollers.NamespaceCache
+	clusterRolesCache rbaccontrollers.ClusterRoleCache
+	rolesCache        rbaccontrollers.RoleCache
+	clusterRoles      rbaccontrollers.ClusterRoleClient
+	roles             rbaccontrollers.RoleClient
+	roleBindingsCache rbaccontrollers.RoleBindingCache
+	roleBindings      rbaccontrollers.RoleBindingController
+	apis              APIResourceWatcher
+}
+
+func (h *rbacHandler) OnChange(_ string, rb *rbacv1.RoleBinding) (*rbacv1.RoleBinding, error) {
```

Comment:
The alternative to this controller is to keep a special dummy namespace with a free-for-all rolebinding and do the SAR check in code. That's still on the table, and one could argue it's much less invasive than duplicating all these rolebindings.

Reply:
Thanks for explaining a bit more, that definitely helps me to understand. If we want to take this route, I think you need 4 total controllers:

2 Binding controllers (RoleBinding and ClusterRoleBinding)
2 Role controllers (Role and ClusterRole).

You need the Role/ClusterRole controllers to handle the case of if a currently bound role changes permissions. For exmample, suppose I have RoleBinding `x` binds user `y` to role `z`. If I change the rules of role `z`, this controller will need to be re-enqueued to update the collections roles with the new rules. 

If you don't want to do a controller for this, you can instead use wrangler's related resource functionality to re-enqueue the binding controller for RoleBindings/ClusterRoleBindings which are attached to this role if you'd prefer that approach instead. The multiple controller approach has it's benefits - namely, you can separate out "which role gives what rule" logic into the role/clusterrole controller and let the bindings just handle bindings, but that's ultimately up to you.

> This is meant to address non-project RBAC, but your comment makes me realize that users could manually create rolebindings in project namespaces too and this would need to work on that, which it currently doesn't.

My initial thought here is that there's a lot of logic here for a more niche use case (users not using our RBAC CRDS in downstream clusters). Though I suppose that we are doing this so that the UI works in all circumstances, and without this the UI could fail to display resources in certain namespaces where they used default k8s rbac, which is probably not great either.

> The alternative to this controller is to keep a special dummy namespace with a free-for-all rolebinding and do the SAR check in code. That's still on the table, and one could argue it's much less invasive than duplicating all these rolebindings.

I don't think I understand this option very well. Would this option be to have a `cattle-unscoped` namespace that contains the `collections.cattle.io` bindings for non-project namespaces? Or would it be one dummy namespace for each real non-project namespace?

Reply:
Forgot to fully address some of these issues before pushing the latest change, will fix.


> I don't think I understand this option very well. Would this option be to have a cattle-unscoped namespace that contains the collections.cattle.io bindings for non-project namespaces? Or would it be one dummy namespace for each real non-project namespace?

The first, one dummy namespace for all non-project namespaces. So a namespace like `cattle-unscoped` would have a rolebinding like

```
subjects:
- kind: Group
  name: system:authenticated
roleRef:
  kind: ClusterRole
  name: cattle-collections-all
```

and a ClusterRole that permits '*' on 'collections.cattle.io'. This lets steve know that there is a "partition" that the user has access to so it can hit /namespaces/cattle-unscoped/{resource}.

> My initial thought here is that there's a lot of logic here for a more niche use case

Agreed, these are for very niche corner cases, but the dashboard doesn't care what is niche or not, it needs to work in all cases.

I think the question is, is it so niche that taking the slight performance penalty of doing this with the SAR check worth the reduction in complexity? With you pointing out that we may need four controllers here I may be leaning toward yes.

Reply:
Resolving, approach changed, file no longer exists.